gobblin-yarn/src/main/java/gobblin/yarn/YarnService.java

```diff
@@ -0,0 +1,457 @@
+    for (String hdfsFilePath : SPLITTER.split(hdfsFileList)) {
+      Path srcFilePath = new Path(hdfsFilePath);
+      YarnHelixUtils.addFileAsLocalResource(
+          srcFilePath.getFileSystem(this.yarnConfiguration), srcFilePath, LocalResourceType.FILE, resourceMap);
+    }
+  }
+
+  private ByteBuffer getSecurityTokens() throws IOException {
+    Credentials credentials = UserGroupInformation.getCurrentUser().getCredentials();
+    Closer closer = Closer.create();
+    try {
+      DataOutputBuffer dataOutputBuffer = closer.register(new DataOutputBuffer());
+      credentials.writeTokenStorageToStream(dataOutputBuffer);
+
+      // Remove the AM->RM token so that containers cannot access it
+      Iterator<Token<?>> tokenIterator = credentials.getAllTokens().iterator();
+      while (tokenIterator.hasNext()) {
+        Token<?> token = tokenIterator.next();
+        if (token.getKind().equals(AMRMTokenIdentifier.KIND_NAME)) {
+          tokenIterator.remove();
+        }
+      }
+
+      return ByteBuffer.wrap(dataOutputBuffer.getData(), 0, dataOutputBuffer.getLength());
+    } catch (Throwable t) {
+      throw closer.rethrow(t);
+    } finally {
+      closer.close();
+    }
+  }
+
+  private String buildContainerCommand(Container container) {
+    String containerProcessName = GobblinWorkUnitRunner.class.getSimpleName();
+    return new StringBuilder()
+        .append(ApplicationConstants.Environment.JAVA_HOME.$()).append("/bin/java")
+        .append(" -Xmx").append(container.getResource().getMemory()).append("M")
+        .append(" ").append(this.containerJvmArgs)
+        .append(" ").append(GobblinWorkUnitRunner.class.getName())
+        .append(" --").append(GobblinYarnConfigurationKeys.APPLICATION_NAME_OPTION_NAME)
+        .append(" ").append(this.applicationName)
+        .append(" 1>").append(ApplicationConstants.LOG_DIR_EXPANSION_VAR).append(File.separator).append(
+          containerProcessName).append(".").append(ApplicationConstants.STDOUT)
+        .append(" 2>").append(ApplicationConstants.LOG_DIR_EXPANSION_VAR).append(File.separator).append(
+          containerProcessName).append(".").append(ApplicationConstants.STDERR)
+        .toString();
+  }
+
+  /**
+   * A custom implementation of {@link AMRMClientAsync.CallbackHandler}.
+   */
+  private class AMRMClientCallbackHandler implements AMRMClientAsync.CallbackHandler {
+
+    private volatile boolean done = false;
+
+    @Override
+    public void onContainersCompleted(List<ContainerStatus> statuses) {
+      for (ContainerStatus containerStatus : statuses) {
+        LOGGER.info(String.format("Container %s has completed with exit status %d",
+            containerStatus.getContainerId(), containerStatus.getExitStatus()));
+        containerMap.remove(containerStatus.getContainerId());
+        containerToParticipantMap.remove(containerStatus.getContainerId());
```

Comment:
Are all the callbacks executed by a single thread? If not we may want to add a `synchronized` block around the two previous lines, so that `containerMap` and `containerToParticipantMap` are always consistent (e.g. they always contain the same `ContainerId`s.

Same goes for all other mutations of `containerMap` and `containerToParticipantMap`.